vowpalwabbit/autolink.cc

```diff
@@ -1,55 +1,75 @@
-#include "reductions.h"
+#include "autolink.h"
 
 using namespace VW::config;
 
-constexpr int autoconstant = 524267083;
+VW::autolink::autolink(uint32_t poly_degree, uint32_t stride_shift)
+  : _poly_degree(poly_degree), _stride_shift(stride_shift)
+{}
 
-struct autolink
+void VW::autolink::predict(LEARNER::single_learner& base, example& ec)
 {
-  uint32_t d;  // degree of the polynomial
-  uint32_t stride_shift;
-};
+  prepare_example(base, ec);
+  base.predict(ec);
+  reset_example(ec);
+}
 
-template <bool is_learn>
-void predict_or_learn(autolink& b, LEARNER::single_learner& base, example& ec)
+void VW::autolink::learn(LEARNER::single_learner& base, example& ec)
+{
+  prepare_example(base, ec);
+  base.learn(ec);
+  reset_example(ec);
+}
+
+void VW::autolink::prepare_example(LEARNER::single_learner& base, example& ec)
 {
   base.predict(ec);
   float base_pred = ec.pred.scalar;
-  // add features of label
+
+  // Add features of label.
   ec.indices.push_back(autolink_namespace);
   features& fs = ec.feature_space[autolink_namespace];
-  for (size_t i = 0; i < b.d; i++)
+  for (size_t i = 0; i < _poly_degree; i++)
+  {
     if (base_pred != 0.)
     {
-      fs.push_back(base_pred, autoconstant + (i << b.stride_shift));
+      fs.push_back(base_pred, AUTOCONSTANT + (i << _stride_shift));
       base_pred *= ec.pred.scalar;
     }
+  }
   ec.total_sum_feat_sq += fs.sum_feat_sq;
+}
 
-  if (is_learn)
-    base.learn(ec);
-  else
-    base.predict(ec);
-
+void VW::autolink::reset_example(example& ec)
+{
+  features& fs = ec.feature_space[autolink_namespace];
   ec.total_sum_feat_sq -= fs.sum_feat_sq;
   fs.clear();
   ec.indices.pop();
 }
 
+template <bool is_learn>
+void predict_or_learn(VW::autolink& b, LEARNER::single_learner& base, example& ec)
+{
+  if (is_learn)
+    b.learn(base, ec);
+  else
+    b.predict(base, ec);
+}
+
 LEARNER::base_learner* autolink_setup(options_i& options, vw& all)
 {
-  free_ptr<autolink> data = scoped_calloc_or_throw<autolink>();
+  uint32_t d;
```

Comment:
This seems like an elegance issue because you need to declare temporary variables and shuffle things into the more permanent structure.   Why not declare the struct in advance?

Reply:
Putting an integer on the stack is significantly cheaper than allocating a reduction object on the heap. It is best to delegate that to when we need it. We can also then make proper use of the constructor and have constant members in the reduction object set in the constructor.

Reply:
Currently we allocate every reduction and for almost all of them delete it right away

Reply:
This is computationally meaningless since it is done once per VW run.  (Or once per pipeline.)  

W.r.t. the constructor, I think it's important to keep things simple for people implementing reductions.  Adding setter/getter functions for every element is not simple.

Reply:
Sure it may be computationally insignificant compared to the hot predict/learn loop, it is still not needed though

Using a constructor gives us lots of wins in terms of guaranteeing state of objects, this gets us out of the setup/finish situation, where an object can be in an indeterminate state.

You'll notice there are no setters or getters in this PR. All of the state is encapsulated in the reduction itself. It allows us to reason about the state of the system in self contained components.

As far as for the people implementing reductions, the trend of changes we are making forces people to think in more of an object based C++ mindset rather than a more C style mindset.

Reply:
W.r.t. 'not needed', we also don't need the extra lines of code.  In general, I think reducing the number of lines of code at negligible computational cost is a good tradeoff.

Using a constructor seems fine, and you can do so here by initializing with d=1.

I agree that having the state encapsulated is great (seems orthogonal to the discussion)?

The general approach of encapsulation which object based C++ provides seems good if we can avoid the downsides.

Reply:
I fail to see why lines of code is an important metric. In my mind speed, safety, correctness and maintainability are what I am trying to achieve. By having more lines here we can construct using the parsed option. This is both faster, and more safe and maintainable because we have a const member.

I'm not sure I follow the comment of initializing d=1. By making d a constructor parameter we can make it const, so that for the rest of the lifetime of that object d will never change.

Reply:
This may be a difference in experience, but doing things without unnecessary shuffling seems more understandable (=maintainable?) to me.

Perhaps we should have a conversation in person in more depth.

Reply:
Yep I agree, we're probably thinking about this differently and aren't matching up 😄 
I'll schedule something for Monday or Tuesday and we can go through it.